// sia-reactor/dist/utils.js
// src/utils/obj.ts
var arrRx = /^([^\[\]]+)\[(\d+)\]$/;
function isDef(val) {
  return val !== void 0;
}
function isArr(obj) {
  return Array.isArray(obj);
}
function isObj(obj, checkArr = true) {
  return "object" === typeof obj && obj !== null && (checkArr ? !Array.isArray(obj) : true);
}
function isStrictObj(obj, crossRealms = false, typecheck = true) {
  return (typecheck ? isObj(obj, false) : true) && (crossRealms ? Object.prototype.toString.call(obj) === "[object Object]" : obj.constructor === Object);
}
function isIter(obj) {
  return obj != null && "function" === typeof obj[Symbol.iterator];
}
function inBoolArrOpt(opt, str) {
  return opt?.includes?.(str) ?? opt;
}
function setAny(target, key, value, separator = ".", keyFunc) {
  if (!key.includes(separator)) return void (target[keyFunc ? keyFunc(key) : key] = value);
  const keys = key.split(separator);
  for (let currObj = target, i = 0, len = keys.length; i < len; i++) {
    const key2 = keyFunc ? keyFunc(keys[i]) : keys[i], match = key2.includes("[") && key2.match(arrRx);
    if (match) {
      const [, key3, iStr] = match;
      if (!isArr(currObj[key3])) currObj[key3] = [];
      if (i === len - 1) currObj[key3][Number(iStr)] = value;
      else currObj[key3][Number(iStr)] ||= {}, currObj = currObj[key3][Number(iStr)];
    } else {
      if (i === len - 1) currObj[key2] = value;
      else currObj[key2] ||= {}, currObj = currObj[key2];
    }
  }
}
function getAny(source, key, separator = ".", keyFunc) {
  if (!key.includes(separator)) return source[keyFunc ? keyFunc(key) : key];
  const keys = key.split(separator);
  let currObj = source;
  for (let i = 0, len = keys.length; i < len; i++) {
    const key2 = keyFunc ? keyFunc(keys[i]) : keys[i], match = key2.includes("[") && key2.match(arrRx);
    if (match) {
      const [, key3, iStr] = match;
      if (!isArr(currObj[key3]) || !(key3 in currObj)) return void 0;
      currObj = currObj[key3][Number(iStr)];
    } else {
      if (!isObj(currObj) || !(key2 in currObj)) return void 0;
      currObj = currObj[key2];
    }
  }
  return currObj;
}
function deleteAny(target, key, separator = ".", keyFunc) {
  if (!key.includes(separator)) return void delete target[keyFunc ? keyFunc(key) : key];
  const keys = key.split(separator);
  for (let currObj = target, i = 0, len = keys.length; i < len; i++) {
    const key2 = keyFunc ? keyFunc(keys[i]) : keys[i], match = key2.includes("[") && key2.match(arrRx);
    if (match) {
      const [, key3, iStr] = match;
      if (!isArr(currObj[key3]) || !(key3 in currObj)) return;
      if (i === len - 1) delete currObj[key3][Number(iStr)];
      else currObj = currObj[key3][Number(iStr)];
    } else {
      if (!isObj(currObj) || !(key2 in currObj)) return;
      if (i === len - 1) delete currObj[key2];
      else currObj = currObj[key2];
    }
  }
}
function inAny(source, key, separator = ".", keyFunc) {
  if (!key.includes(separator)) return key in source;
  const keys = key.split(separator);
  for (let currObj = source, i = 0, len = keys.length; i < len; i++) {
    const key2 = keyFunc ? keyFunc(keys[i]) : keys[i], match = key2.includes("[") && key2.match(arrRx);
    if (match) {
      const [, key3, iStr] = match;
      if (!isArr(currObj[key3]) || !(key3 in currObj)) return false;
      if (i === len - 1) return true;
      currObj = currObj[key3][Number(iStr)];
    } else {
      if (!isObj(currObj) || !(key2 in currObj)) return false;
      if (i === len - 1) return true;
      currObj = currObj[key2];
    }
  }
  return true;
}
function parseAnyObj(obj, separator = ".", keyFunc = (p) => p, visited = /* @__PURE__ */ new WeakSet()) {
  if (!isObj(obj) || visited.has(obj)) return obj;
  visited.add(obj);
  const result = {};
  Object.keys(obj).forEach((k) => k.includes(separator) ? setAny(result, k, parseAnyObj(obj[k], separator, keyFunc, visited), separator, keyFunc) : result[k] = isObj(obj[k]) ? parseAnyObj(obj[k], separator, keyFunc, visited) : obj[k]);
  return result;
}
function parseEvOpts(options, opts, boolOpt = opts[0], result = {}) {
  return Object.assign(result, "boolean" === typeof options ? { [boolOpt]: options } : options), result;
}
function mergeObjs(o1 = {}, o2 = {}) {
  const merged = { ...o1 || {}, ...o2 || {} };
  return Object.keys(merged).forEach((k) => isObj(o1?.[k]) && isObj(o2?.[k]) && (merged[k] = mergeObjs(o1[k], o2[k]))), merged;
}
function getTrailPaths(path, reverse = true) {
  const parts = path.split("."), chain = ["*"];
  let acc = "";
  for (let i = 0, len = parts.length; i < len; i++) chain.push(acc += (i === 0 ? "" : ".") + parts[i]);
  return reverse ? chain.reverse() : chain;
}
function getTrailRecords(obj, path) {
  const parts = path.split("."), record = [["*", obj, obj]];
  let acc = "", currObj = obj;
  for (let i = 0, len = parts.length; i < len; i++) record.push([acc += (i ? "." : "") + parts[i], currObj, currObj = currObj?.[parts[i]]]);
  return record;
}
function deepClone(obj, crossRealms, visited = /* @__PURE__ */ new WeakMap()) {
  if (!(isStrictObj(obj, crossRealms) || isArr(obj)) || visited.has(obj)) return obj;
  const clone = isArr(obj) ? [] : {};
  visited.set(obj, clone);
  const keys = Object.keys(obj);
  for (let i = 0, len = keys.length; i < len; i++) clone[keys[i]] = deepClone(obj[keys[i]], crossRealms, visited);
  return clone;
}
function nuke(target) {
  let proto = target;
  while (proto && proto !== Object.prototype) {
    const keys = Object.getOwnPropertyNames(proto);
    for (let i = 0, len = keys.length; i < len; i++) {
      if (keys[i] === "constructor") continue;
      const desc = Object.getOwnPropertyDescriptor(proto, keys[i]);
      if (desc && ("function" === typeof desc.value || desc.get || desc.set)) continue;
      proto[keys[i]] = null;
    }
    proto = Object.getPrototypeOf(proto);
  }
}
export {
  deepClone,
  deleteAny,
  getAny,
  getTrailPaths,
  getTrailRecords,
  inAny,
  inBoolArrOpt,
  isArr,
  isDef,
  isIter,
  isObj,
  isStrictObj,
  mergeObjs,
  nuke,
  parseAnyObj,
  parseEvOpts,
  setAny
};
