// @t007/utils/dist/index.js
// src/core/str.ts
function uid(prefix = "") {
  return prefix + Date.now().toString(36) + "_" + performance.now().toString(36).replace(".", "") + "_" + Math.random().toString(36).slice(2);
}
function isSameURL(src1, src2) {
  if (typeof src1 !== "string" || typeof src2 !== "string" || !src1 || !src2) return false;
  try {
    const u1 = new URL(src1, window.location.href), u2 = new URL(src2, window.location.href);
    return decodeURIComponent(u1.origin + u1.pathname) === decodeURIComponent(u2.origin + u2.pathname);
  } catch {
    return src1.replace(/\\/g, "/").split("?")[0].trim() === src2.replace(/\\/g, "/").split("?")[0].trim();
  }
}

// src/core/dom.ts
function createEl(tag, props, dataset, styles, el = tag ? document?.createElement(tag) : null) {
  return assignEl(el, props, dataset, styles), el;
}
function assignEl(el, props, dataset, styles) {
  if (!el) return;
  if (props) {
    for (const k of Object.keys(props)) if (props[k] !== void 0) el[k] = props[k];
  }
  if (dataset) {
    for (const k of Object.keys(dataset)) if (dataset[k] !== void 0) el.dataset[k] = String(dataset[k]);
  }
  if (styles) {
    for (const k of Object.keys(styles)) if (styles[k] !== void 0) el.style[k] = styles[k];
  }
}
var VIRTUAL_RESOURCE = /* @__PURE__ */ Symbol.for("T007_VIRTUAL_RESOURCE");
function loadResource(req, type = "style", { module, media, crossOrigin, integrity, referrerPolicy, nonce, fetchPriority, attempts = 3, retryKey = false } = {}, w = window) {
  w.t007 ??= {}, w.t007._resourceCache ??= {};
  if (req === VIRTUAL_RESOURCE || "symbol" === typeof req) return Promise.resolve();
  const src = req;
  if (w.t007._resourceCache[src]) return w.t007._resourceCache[src];
  const existing = type === "script" ? Array.prototype.find.call(w.document.scripts, (s) => isSameURL(s.src, src)) : type === "style" ? Array.prototype.find.call(w.document.styleSheets, (s) => isSameURL(s.href, src)) : null;
  if (existing) return w.t007._resourceCache[src] = Promise.resolve(existing);
  w.t007._resourceCache[src] = new Promise((resolve, reject) => {
    (function tryLoad(remaining, el) {
      const onerror = () => {
        el?.remove?.();
        if (remaining > 1) {
          setTimeout(tryLoad, 1e3, remaining - 1);
          console.warn(`Retrying ${type} load (${attempts - remaining + 1}): ${src}...`);
        } else {
          delete w.t007._resourceCache[src];
          reject(new Error(`${type} load failed after ${attempts} attempts: ${src}`));
        }
      };
      const url = retryKey && remaining < attempts ? `${src}${src.includes("?") ? "&" : "?"}_${retryKey}=${Date.now()}` : src;
      if (type === "script") w.document.body.append(el = createEl("script", { src: url, type: module ? "module" : "text/javascript", crossOrigin, integrity, referrerPolicy, nonce, fetchPriority, onload: () => resolve(el), onerror }) || "");
      else if (type === "style") w.document.head.append(el = createEl("link", { rel: "stylesheet", href: url, media, crossOrigin, integrity, referrerPolicy, nonce, fetchPriority, onload: () => resolve(el), onerror }) || "");
      else reject(new Error(`Unsupported resource type: ${type}`));
    })(attempts);
  });
  return w.t007._resourceCache[src];
}

// src/core/num.ts
function clamp(min = 0, val, max = Infinity) {
  return Math.min(Math.max(val, min), max);
}

// src/mixins/methodist.ts
function onAllMethods(owner, callback) {
  let proto = owner;
  while (proto && proto !== Object.prototype) {
    for (const method of Object.getOwnPropertyNames(proto)) {
      if (method === "constructor") continue;
      if ("function" !== typeof Object.getOwnPropertyDescriptor(proto, method)?.value) continue;
      callback(method, owner);
    }
    proto = Object.getPrototypeOf(proto);
  }
}
function bindAllMethods(owner) {
  onAllMethods(owner, (method, owner2) => {
    owner2[method] = owner2[method].bind(owner2);
  });
}
function guardAllMethods(owner, guardFn = guardMethod, bound = true) {
  onAllMethods(owner, (method, owner2) => {
    owner2[method] = guardFn(bound ? owner2[method].bind(owner2) : owner2[method]);
  });
}
function guardMethod(fn, onError = (e) => console.error(e)) {
  return ((...args) => {
    try {
      const result = fn(...args);
      return result instanceof Promise ? result.catch((e) => onError(e)) : result;
    } catch (e) {
      onError(e);
    }
  });
}

// src/quirks/scroll.ts
function initVScrollerator({ baseSpeed = 3, maxSpeed = 10, stepDelay = 2e3, baseRate = 16, lineHeight = 80, margin = 80, car = window } = {}) {
  let linesPerSec = baseSpeed, accelId = null, lastTime = null;
  const drive = (clientY, brake = false, offsetY = 0) => {
    if (car !== window) clientY -= offsetY;
    const now = performance.now(), speed = linesPerSec * lineHeight * ((lastTime ? now - lastTime : baseRate) / 1e3);
    if (!brake && (clientY < margin || clientY > (car.innerHeight ?? car.offsetHeight) - margin)) {
      accelId === null ? accelId = setTimeout(() => linesPerSec += 1, stepDelay) : linesPerSec > baseSpeed && (linesPerSec = Math.min(linesPerSec + 1, maxSpeed));
      car.scrollBy?.(0, clientY < margin ? -speed : speed);
    } else reset();
    return lastTime = !brake ? now : null, speed;
  };
  const reset = () => (accelId && clearTimeout(accelId), accelId = null, linesPerSec = baseSpeed, lastTime = null);
  return { drive, reset };
}
function initScrollAssist(el, { pxPerSecond = 80, assistClassName = "tmg-video-controls-scroll-assist", vertical = true, horizontal = true } = {}) {
  t007._scrollers ??= /* @__PURE__ */ new WeakMap();
  t007._scroller_r_observer ??= new ResizeObserver((entries) => entries.forEach(({ target }) => t007._scrollers.get(target)?.update()));
  t007._scroller_m_observer ??= new MutationObserver((entries) => {
    const els = /* @__PURE__ */ new Set();
    for (const entry of entries) {
      let node = entry.target instanceof Element ? entry.target : null;
      while (node && !t007._scrollers.has(node)) node = node.parentElement;
      if (node) els.add(node);
    }
    for (const el2 of els) t007._scrollers.get(el2)?.update();
  });
  const parent = el?.parentElement;
  if (!parent || t007._scrollers.has(el)) return;
  const assist = {};
  let scrollId = null, last = performance.now(), assistWidth = 20, assistHeight = 20;
  const update = () => {
    const hasInteractive = !!parent.querySelector('button, a[href], input, select, textarea, [contenteditable="true"], [tabindex]:not([tabindex="-1"])');
    if (horizontal) {
      const w = assist.left?.offsetWidth || assistWidth, check = hasInteractive ? el.clientWidth < w * 2 : false;
      assist.left.style.display = check ? "none" : el.scrollLeft > 0 ? "block" : "none";
      assist.right.style.display = check ? "none" : el.scrollLeft + el.clientWidth < el.scrollWidth - 1 ? "block" : "none";
      assistWidth = w;
    }
    if (vertical) {
      const h = assist.up?.offsetHeight || assistHeight, check = hasInteractive ? el.clientHeight < h * 2 : false;
      assist.up.style.display = check ? "none" : el.scrollTop > 0 ? "block" : "none";
      assist.down.style.display = check ? "none" : el.scrollTop + el.clientHeight < el.scrollHeight - 1 ? "block" : "none";
      assistHeight = h;
    }
  };
  const scroll = (dir) => {
    const frame = () => {
      const now = performance.now(), dt = now - last;
      last = now;
      const d = pxPerSecond * dt / 1e3;
      if (dir === "left") el.scrollLeft = Math.max(0, el.scrollLeft - d);
      if (dir === "right") el.scrollLeft = Math.min(el.scrollWidth - el.clientWidth, el.scrollLeft + d);
      if (dir === "up") el.scrollTop = Math.max(0, el.scrollTop - d);
      if (dir === "down") el.scrollTop = Math.min(el.scrollHeight - el.clientHeight, el.scrollTop + d);
      scrollId = requestAnimationFrame(frame);
    };
    last = performance.now();
    frame();
  };
  const stop = () => (cancelAnimationFrame(scrollId ?? 0), scrollId = null);
  const addAssist = (dir) => {
    const div = createEl("div", { className: assistClassName }, { scrollDirection: dir }, { display: "none" });
    if (!div) return;
    ["pointerenter", "dragenter"].forEach((evt) => div.addEventListener(evt, () => scroll(dir)));
    ["pointerleave", "pointerup", "pointercancel", "dragleave", "dragend"].forEach((evt) => div.addEventListener(evt, stop));
    dir === "left" || dir === "up" ? parent.insertBefore(div, el) : parent.append(div);
    assist[dir] = div;
  };
  if (horizontal) ["left", "right"].forEach(addAssist);
  if (vertical) ["up", "down"].forEach(addAssist);
  el.addEventListener("scroll", update);
  t007._scroller_r_observer.observe(el);
  t007._scroller_m_observer.observe(el, { childList: true, subtree: true, characterData: true });
  t007._scrollers.set(el, {
    update,
    destroy() {
      stop();
      el.removeEventListener("scroll", update);
      t007._scroller_r_observer.unobserve(el);
      t007._scrollers.delete(el);
      Object.values(assist).forEach((a) => a.remove());
    }
  });
  return update(), t007._scrollers.get(el);
}
var removeScrollAssist = (el) => t007._scrollers.get(el)?.destroy();

// src/index.ts
if (typeof window !== "undefined") {
  window.t007 ??= {};
  t007.VIRTUAL_RESOURCE = VIRTUAL_RESOURCE;
  window.T007_TOAST_JS_SRC ??= `https://cdn.jsdelivr.net/npm/@t007/toast@latest`;
  window.T007_INPUT_JS_SRC ??= `https://cdn.jsdelivr.net/npm/@t007/input@latest`;
  window.T007_DIALOG_JS_SRC ??= `https://cdn.jsdelivr.net/npm/@t007/dialog@latest`;
  window.T007_TOAST_CSS_SRC ??= `https://cdn.jsdelivr.net/npm/@t007/toast@latest/dist/index.min.css`;
  window.T007_INPUT_CSS_SRC ??= `https://cdn.jsdelivr.net/npm/@t007/input@latest/dist/index.min.css`;
  window.T007_DIALOG_CSS_SRC ??= `https://cdn.jsdelivr.net/npm/@t007/dialog@latest/dist/index.min.css`;
}
export {
  VIRTUAL_RESOURCE,
  assignEl,
  bindAllMethods,
  clamp,
  createEl,
  guardAllMethods,
  guardMethod,
  initScrollAssist,
  initVScrollerator,
  isSameURL,
  loadResource,
  onAllMethods,
  removeScrollAssist,
  uid
};
